### ISSUE TITLE
Cannot replicate 97.64 on megaface using the released LResNet50E-IR model

### ISSUE THREAD

**Issue by juicecwc** [deepinsight/insightface#167]
I have tried many times and only got 96.58 on megaface, does anyone else have replicated this result? In my case, the first 10 number of Adam_Brody_420.png's feature is [ 0.0424159 ,  0.05080357,  0.00801507, -0.04773382,  0.00920391, 0.03691193, -0.01137877, -0.02552283,  0.01239266,  0.00478121], can u tell me whether it is right or not? Thank you! @nttstar

**Comment by nttstar** [comment]
Did you download the pre-aligned facescrub dataset from here?

**Comment by juicecwc** [comment]
No, I align facescrub use your code by myself. But I didn't find any facescrub dataset in this repo.

**Comment by nttstar** [comment]
https://pan.baidu.com/s/1ghcpIH9

**Comment by juicecwc** [comment]
Thank you, I'll try soon! What other processing have you done to facescrub?

**Comment by meanmee** [comment]
I doenload the facescrub, but just got 0.949282

**Comment by bruinxiong** [comment]
@juicecwc Did you train your own ArcFace model with ms1m dataset given by @nttstar ? Which epoch is selected for feature extraction ? How did you preprocess the megaface and facescrub dataset ? I have trained ArcFace r100 model with ms1m given by @nttstar, during training stage, the test performance of LFW(given by @nttstar as bin file) is achieved 99.8+, I choose the model with that epoch (92), but the megaface performance is very low, only 42.7. I use our own preprocessing method on megaface and facescrub and obtain the image with 112x112. could you mind to share the feature extraction code for megaface evaluation ? I checked my code, there is no bug. However, the performance is quite low.

**Comment by staceycy** [comment]
@meanmee Hi, could you please tell how did you processed the MegaFace dataset? Did you used the align_megaface.py script provided in the src/align/ folder ? Thank you.

**Comment by meanmee** [comment]
Yes， I use the align_megaface.py script
@staceycy

**Comment by staceycy** [comment]
Noted. Thank you very much. @meanmee 
When we changed aligned MegaFace from our own version to that processed with align_megaface.py, the performance become normal. 
Data processing does affects a lot on model performance.

**Comment by meanmee** [comment]
Could you tell me about what you changed in align_megaface.py？

**Comment by bruinxiong** [comment]
@nttstar @meanmee @juicecwc 
Firstly, thanks @nttstar  insightface provides this source code to attract many face researchers.
Secondly, I and @staceycy  train LResNet100E-IR on MS1M dataset s(given by insightface) with ArcFace Loss and evaluate it on Megaface Challenge 1. We make some comparison with different data pre-processing. First, we follow align_megaface.py to detect and align megaface distractor dataset and utilized facescrub 112x112 dataset directly given by insightface. The performance is listed as follow: **Identification Rate vs. Rank 97.766%, Verification 98.693%**. Second, we follow align_megaface.py to detect and align megaface distractor dataset and we also follow align_facescrub.py to detect and align facescrub dataset. The performance is listed as follow: **Identification Rate vs. Rank 96.729%, Verification 97.509%**. It's obvious that there is a gap between them. In order to verify the difference between facescrub 112x112 given by insightface and our reproduced facescrub images, we show them as the picture, we can find there is some difference between them. **Based on comparison with many faces, basically, the difference may visually come from affine transformation.** 
![comparison](https://user-images.githubusercontent.com/3366247/39609781-2173136e-4f7e-11e8-88cc-2ec04dffbee9.png)
Finally, we have two questions, 
1. Does insightface utilizes align_facescrub.py to pre-process the facescrub dataset and align_megaface.py to pre-process the megaface distractor dataset ?
2. Does insightface utilizes align_insight.py or align_dataset_mtcnn.py in src/align/ to pre-process the training data such as MS1M and VGG2 ?

Thanks!

**Comment by staceycy** [comment]
Hi, sorry for late reply. @meanmee 
I didn't change the align_megaface.py, I just used the provided one. 
Previously, we get strange evaluation result due to different data processing methods.
First, we used our own way to align both megaface and facescrub, the result was quite low. 
Then, we used the aligned facescrub provided by @nttstar and aligned megaface with our own method, the result was extremely high, this is because align megaface and facescrub in different way will differentiate the two datasets which boot the performance extremely high. 
Then, we do as @bruinxiong said above. 
We realized that the way to align face is quite important so we are wondering to know how @nttstar align both the training and testing data. 
Thank you all for your contribution and discussion.

**Comment by yangfly** [comment]
@nttstar @bruinxiong How do you get `Verification` or `VR@FAR10^6` score on megaface?
I run command like:
```shell
python run_experiment.py -s 1000000 \
    path_megaface_feature path_facescrub_feature _arcface.bin path_results/
```
And I got output with pretrained-R50E (cleaned)
```log
Done matching! Score matrix size: 3530 1000000
Saving to /home/data/insight/megaface/raw_eval/r50/results_cm/otherFiles/facescrub_megaface_r50_112x112_1000000_1.bin
Computing test results with 1000000 images for set 1
Loaded 3530 probes spanning 80 classes
Loading from /home/data/insight/megaface/raw_eval/r50/results_cm/otherFiles/facescrub_facescrub_r50_112x112.bin
Probe score matrix size: 3530 3530
distractor score matrix size: 3530 1000000
Done loading. Time to compute some stats!
Finding top distractors!
Done sorting distractor scores
Making gallery!
Done Making Gallery!
Allocating ranks (1000080)
Rank 1: 0.965815
```
I carefully check all output log, and I couldn't find anything seems to be `Verification` score. 
Thank you.

**Comment by FafeiXuan** [comment]
@meanmee Hi, I want to test the downloaded model r50 on megaface, and I follow the instructions, I changed the path of lst and others, and then run align_megaface.py to align facescrub and megaface, and also run gen_megaface.py and remove_noises.py, and at last I use run_experiment.py, but the result is rank-1 41%, there must be something wrong, what maybe I miss? Thank you!
And I also download facescrub_images_112x112.tar but the pictures are all png, should I change the format as the facescrub_uncropped_features_list.json so that I can run run_experiment.py?
Thanks! Waiting for you :)

**Comment by HaoLiuHust** [comment]
any progress?

**Comment by twmht** [comment]
@bruinxiong 

Do you have any update on this?

**Comment by 2151330** [comment]
Where did you find run_experiment.py?I can't find it?
